### ISSUE TITLE
[Bug] Type inference not working with t.Transform

### ISSUE THREAD

**Issue by mnpenner** [elysiajs/elysia#320]
I wrote a custom type:

```ts
export const ExternalId = t.Transform(
        t.RegExp(ID_REGEX, {
            default: encodeId(0),
        })
    )
    .Decode(decodeId)
    .Encode(encodeId)
```

Where `ExternalId` is inferred as 

```
export const ExternalId: TTransform<TString, number>
```

But when I use it in Elysia with my routes,

```ts
{
        params: t.Object({
            playId: ExternalId
        })
    }
```

Then `params.playId` is inferred to be a `string`.

I think this is a bug because this test passes and doesn't show any TS errors:

```ts
describe('ExternalId', () => {
    const RequestBody = t.Object({
        player: t.Union([t.Literal("__human__"), ExternalId]),
        opponent: ExternalId,
    })

    it('decodes', () => {
        expect(Value.Decode<typeof RequestBody>(RequestBody, {player: encodeId(1), opponent: encodeId(2)})).toEqual({
            player: 1,
            opponent: 2,
        })
    })
})
```

Also, I see in the source code:

https://github.com/elysiajs/elysia/blob/f9587a78b357ddac0d2aa236469392fbfcf5a202/src/type-system.ts#L201

That ` as any as TNumber` looks like a workaround; it fixes `Value.Decode` but breaks `Value.Create`.

**Comment by cztomsik** [comment]
I think there is bigger issue somewhere.

I got around the type inference with `as Xxx`, in my case I wanted to accept date in iso string and read it parsed.

```
const DateTime = t
  .Transform(t.String())
  .Decode(v => new Date(v))
  .Encode(v => v.toISOString()) as any as ReturnType<typeof t.Date>
```

This works, but it does **not work together with `t.Nullable()`** which is weird. It seems it somehow skips the transform entirely.

**Comment by ehaynes99** [comment]
Since TypeBox implemented transformable types, there are 2 new types, `StaticEncode` (which is just an alias for `Static`) represents the over-the-wire type, and `StaticDecode` is decoded type.

@sinclairzx81 explains it here: https://github.com/sinclairzx81/typebox/issues/343#issuecomment-1663275670

The fix would be something like this: https://github.com/ehaynes99/elysia/commit/995165d4aa4ebb2bcf7147d3d8c1cb2e97d8ea87

However, that runs into this issue, and I don't know how to solve it here: https://github.com/sinclairzx81/typebox/issues/680
```
$ bun run build
$ rimraf dist && bun run build:esm && bun run build:bun && bun run build:cjs
$ swc src -d dist && tsc --project tsconfig.esm.json
Successfully compiled: 13 files with swc (33.56ms)
src/index.ts:2101:4 - error TS2589: Type instantiation is excessively deep and possibly infinite.

2101    this.all(
        ~~~~~~~~~
2102     '/',
     ~~~~~~~~
 ... 
2106     } as any
     ~~~~~~~~~~~~
2107    )
     ~~~~
```

**Comment by SaltyAom** [comment]
If I understand correctly, this should have been fixed by using `StaticDecode` instead of `Decode` in 550a012, published under 1.1.0.

Feel free to reopen the issue if this doesn't fix your issue, thanks.
